File: py_solutions_11-20/Euler_19.py
# find the number sundays that fell on the first day of the month in between 
# 1901 - 2000. 

from __future__ import print_function

import timeit
from itertools import cycle

try:
    range = xrange
except NameError:
    pass

start = timeit.default_timer()

def euler_19(start, stop):
    leap_yr = range(1, 30)
    weekday = cycle(range(7))
    for i in range(start, stop):
        Feb = range(1, 29) if i % 4 != 0 else leap_yr
        Jan = Mar = May = Jul = Aug = Oct = Dec = range(1, 32)    
        Apr = Jun = Sep = Nov = range(1, 31) 
        year = [Jan, Feb, Mar, Apr, May, Jun, Jul, Aug, Sep, Oct, Nov, Dec] 
        yield sum([1 for mos in range(12) if 
                   list(zip(year[mos], weekday))[0] == (1, 6)])


print("Answer: {}".format(sum([sunday for sunday in euler_19(1901, 2000)])))
stop = timeit.default_timer()
print("Time: {0:9.5f}".format(stop - start))

File: C/c_solutions_11-20/Euler_19.c
// Find the number of sundays that fell on the first day of the month in 
// between 1901 - 2000

#include "euler_util.h"


int Months[12] = {31, 00, 31, 30, 31, 30,
                  31, 31, 30, 31, 30, 31};

int main(int argc, char *argv[])
{
    float start = timeit();

    int answer;
    int year, mos, day, day_num, duration;

    for (answer=0, day_num=0, year=1901; year < 2000; year++) {
        for (mos=0; mos < 12; mos++) {
            if (mos == 1)
                duration = year % 4 == 0 ? 29 : 28;
            else
                duration = Months[mos];

            for (day=0; day <= duration; day++, day_num++)
                if (day == 0 && day_num % 7 == 0)
                    answer++;
        }
    }

    printf("Answer: %d\n", answer);
    float stop = timeit();
    printf("Time: %f\n", stop - start);

    return 0;
}
